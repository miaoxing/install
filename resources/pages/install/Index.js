/**
 * @layout false
 */
import React from "react";
import {Button, Checkbox, Divider, Form, Input} from "antd";
import {Box, Heading, Image} from 'rebass/styled-components';
import $ from 'miaoxing';
import app from '@miaoxing/app';
import logo from 'plugins/admin/resources/images/logo.png';
import {createGlobalStyle} from 'styled-components';
import api from '@miaoxing/api';
import {AForm, AFormItem} from '@miaoxing/form';

const GlobalStyle = createGlobalStyle`
  body {
    background: #f5f8fa url(https://image-10001577.image.myqcloud.com/uploads/3/20190602/15594729401485.jpg);
    background-size: cover;
  }
`;

export default class extends React.Component {
  async componentDidMount() {
    const ret = await api.curPath('installed');
    if (ret.code !== 1) {
      $.alert(ret.message);
    }
  }

  showAgreement = (e) => {
    alert('xx');
    e.preventDefault();
  }

  handleSubmit = async (values) => {
    const ret = await api.curCreate({
      data: values,
      loading: true,
    });
    if (ret.code !== 1) {
      $.alert(ret.message);
      return ;
    }

    await $.ret(ret);
    window.location = ret.next;
  }

  render() {
    return <>
      <GlobalStyle/>
      <Box
        width={700}
        mx="auto"
        mt={5}
        p={5}
        bg="white"
      >
        <Box
          mb={4}
          textAlign="center"
        >
          <Image height={30} src={logo}/>
        </Box>
        <Heading
          mb={5}
          textAlign="center"
          fontSize={3}
          fontWeight="normal"
          color="muted"
        >
          安装
        </Heading>
        <AForm
          labelCol={{span: 8}}
          wrapperCol={{span: 8}}
          validateMessages={{
            required: '该项是必填的',
          }}
          initialValues={{
            dbHost: 'localhost',
            dbDbName: 'miaoxing',
            dbUser: 'root',
            dbTablePrefix: 'mx_',
            username: 'admin',

            //
            dbHost: 'mysql',
            dbDbName: 'm' + parseInt(Math.random() * 10000, 10),
            dbPassword: 'password',
            password: 'password',
            agree: true,
          }}
          onFinish={this.handleSubmit}
        >
          <AFormItem label="数据库地址" name="dbHost" rules={[{required: true}]}/>
          <AFormItem label="数据库名称" name="dbDbName" rules={[{required: true}]}/>
          <AFormItem label="数据库用户名" name="dbUser" rules={[{required: true}]}/>
          <AFormItem label="数据库密码" name="dbPassword" rules={[{required: true}]}/>
          <AFormItem label="数据表前缀" name="dbTablePrefix" rules={[{required: true}]}/>

          <Divider/>

          <AFormItem label="管理员用户名" name="username" rules={[{required: true}]}/>
          <AFormItem label="管理员密码" name="password" rules={[{required: true}]}/>

          <Form.Item
            name="agree"
            wrapperCol={{offset: 8, span: 8}}
            valuePropName="checked"
            rules={[{required: true, message: "请阅读并同意《服务协议》"}]}
          >
            <Checkbox>同意<a href="#" onClick={this.showAgreement}>《服务协议》</a></Checkbox>
          </Form.Item>

          <Form.Item
            wrapperCol={{offset: 8, span: 8}}
          >
            <Button type="primary" htmlType="submit" block>
              安装
            </Button>
          </Form.Item>
        </AForm>
      </Box>
    </>;
  }
}
