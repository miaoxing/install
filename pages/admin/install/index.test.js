import Index from './index';
import $, {Ret} from 'miaoxing';
import {fireEvent, render, waitFor} from '@testing-library/react';
import {bootstrap, resetUrl, setUrl} from '@mxjs/test';
import {app} from '@mxjs/app';

bootstrap();

describe('install', () => {
  beforeEach(() => {
    setUrl('admin/install');
    app.page = {
      collection: 'admin/install',
      index: true,
    };
  });

  afterEach(() => {
    resetUrl();
    app.page = {};
  });

  test('install', async () => {
    $.http = jest.fn()
      .mockResolvedValueOnce({
        ret: Ret.suc({
          data: {
            installRet: Ret.suc({
              data: [
                Ret.suc('check suc 1'),
                Ret.suc('check suc 2'),
              ],
            }),
            license: 'xxx',
          },
        }),
      })
      // 测试 rewrite
      .mockResolvedValueOnce({
        ret: Ret.suc(),
      })
      .mockResolvedValueOnce({
        ret: Ret.suc(),
      });

    const result = render(<Index/>);

    await result.findByText('check suc 1');

    fireEvent.click(result.getByText('进入安装'));
    fireEvent.change(result.getByLabelText('数据库密码'), {target: {value: 'password'}});
    fireEvent.change(result.getByLabelText('管理员密码'), {target: {value: '123456'}});
    fireEvent.click(result.getByLabelText('安装演示数据'));
    fireEvent.click(result.getByText('我已阅读并同意'));
    fireEvent.click(result.getByText('安 装'));

    await waitFor(() => {
      expect($.http).toBeCalledTimes(3);
    });

    expect($.http).toMatchSnapshot();
  });

  test('check reqs', async () => {
    $.http = jest.fn()
      .mockResolvedValueOnce({
        ret: Ret.suc({
          data: {
            installRet: Ret.err({
              data: [
                Ret.suc('check suc 1'),
                Ret.err('check fail 2'),
              ],
            }),
            license: 'xxx',
          },
        }),
      })
      // 测试 rewrite
      .mockResolvedValueOnce({
        ret: Ret.suc(),
      })
      .mockResolvedValueOnce({
        ret: Ret.suc({
          data: {
            installRet: Ret.err({
              data: [
                Ret.err('check fail 1'),
                Ret.suc('check suc 2'),
              ],
            }),
            license: 'xxx',
          },
        }),
      })
      .mockResolvedValueOnce({
        ret: Ret.suc({
          data: {
            installRet: Ret.suc({
              data: [
                Ret.suc('check suc 1'),
                Ret.suc('check suc 2'),
              ],
            }),
            license: 'xxx',
          },
        }),
      });

    const result = render(<Index/>);

    await result.findByText('check suc 1');

    fireEvent.click(result.getByText('重新检查'));

    await result.findByText('check fail 1');

    fireEvent.click(result.getByText('重新检查'));

    await result.findByText('check suc 1');

    expect(result.queryByText('进入安装')).not.toBeNull();

    expect($.http).toMatchSnapshot();
  });

  test('check install fail', async () => {
    $.http = jest.fn().mockRejectedValueOnce('500');
    $.alert = jest.fn();

    render(<Index/>);

    await waitFor(() => {
      expect($.alert).toBeCalled();
    });

    expect($.http).toMatchSnapshot();
    expect($.alert).toMatchSnapshot();
  });

  test('submit fail', async () => {
    $.http = jest.fn()
      .mockResolvedValueOnce({
        ret: Ret.suc({
          data: {
            installRet: Ret.suc({
              data: [
                Ret.suc('check suc 1'),
                Ret.suc('check suc 2'),
              ],
            }),
            license: 'xxx',
          },
        }),
      })
      // 测试 rewrite
      .mockResolvedValueOnce({
        ret: Ret.suc(),
      })
      .mockRejectedValue(500);

    const result = render(<Index/>);

    const btn = await result.findByText('进入安装');
    fireEvent.click(btn);
    fireEvent.change(result.getByLabelText('数据库密码'), {target: {value: 'password'}});
    fireEvent.change(result.getByLabelText('管理员密码'), {target: {value: '123456'}});
    fireEvent.click(result.getByText('我已阅读并同意'));

    $.alert = jest.fn();

    fireEvent.click(result.getByText('安 装'));

    await waitFor(() => {
      expect($.http).toBeCalledTimes(3);
    });

    expect($.http).toMatchSnapshot();
    // 抛出异常，未执行 alert 逻辑
    expect($.alert).not.toBeCalled();
  });
});
