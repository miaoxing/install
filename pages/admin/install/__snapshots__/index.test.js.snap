// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`install check reqs 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "method": "get",
        "url": "/admin-api/install",
      },
    ],
    Array [
      Object {
        "ignoreError": true,
        "method": "get",
        "url": "../admin-api/install",
      },
    ],
    Array [
      Object {
        "method": "get",
        "url": "/admin-api/install",
      },
    ],
    Array [
      Object {
        "method": "get",
        "url": "/admin-api/install",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {},
    },
    Object {
      "type": "return",
      "value": Promise {},
    },
    Object {
      "type": "return",
      "value": Promise {},
    },
    Object {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`install install 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "method": "get",
        "url": "/admin-api/install",
      },
    ],
    Array [
      Object {
        "ignoreError": true,
        "method": "get",
        "url": "../admin-api/install",
      },
    ],
    Array [
      Object {
        "data": Object {
          "agree": true,
          "dbDbName": "miaoxing",
          "dbHost": "localhost",
          "dbPassword": "password",
          "dbTablePrefix": "mx_",
          "dbUser": "root",
          "password": "123456",
          "requestDefaultUrlRewrite": true,
          "seed": true,
          "username": "admin",
        },
        "loading": true,
        "method": "post",
        "url": "/admin-api/install",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {},
    },
    Object {
      "type": "return",
      "value": Promise {},
    },
    Object {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;
